Sources/Jobs/Scheduler/DateExtensions/DateExtensionManipulation.swift

```diff
@@ -0,0 +1,102 @@
+import Foundation
+
+extension Date {
+    enum DateManipulationError: Error {
+        case couldNotFindNextDate
+        case couldNotValidateDateComponentValue
+    }
+
+    func setValuesToDefault(LowerThan ruleTimeUnit: RecurrenceRuleTimeUnit, date: Date) -> Date {
+        var lowerCadenceLevelUnits = [RecurrenceRuleTimeUnit]()
+        switch ruleTimeUnit {
+        case .second:
+            lowerCadenceLevelUnits = []
+        case .minute:
+            lowerCadenceLevelUnits = [.second]
+        case .hour:
+            lowerCadenceLevelUnits = [.second, .minute]
+        case .dayOfMonth, .dayOfWeek, .weekOfMonth, .weekOfYear:
+            lowerCadenceLevelUnits = [.second, .minute, .hour]
+        case .month:
+            lowerCadenceLevelUnits = [.second, .minute, .hour, .dayOfMonth]
+        case .quarter, .year:
+            lowerCadenceLevelUnits = [.second, .minute, .hour, .dayOfMonth, .month]
+        }
+
+        var newDate = date
+        for unit in lowerCadenceLevelUnits {
+            do {
+                newDate = try setValueToDefault(unit, date: newDate)
+            } catch {
+                // do nothing
+            }
+        }
+
+        return newDate
+    }
+
+    func setValueToDefault(_ ruleTimeUnit: RecurrenceRuleTimeUnit, date: Date) throws -> Date {
+        guard let defaultValue = resolveValidLowerBound(ruleTimeUnit) else {
+            return date
+        }
+        guard let currentValue = self.resolveDateComponentValue(for: ruleTimeUnit) else {
+            throw DateManipulationError.couldNotValidateDateComponentValue
+        }
+
+        let dateComponent = resolveCalendarComponent(for: ruleTimeUnit)
+        let unitsToSubtract = currentValue - defaultValue
+
+        return Calendar.current.date(byAdding: dateComponent, value: -unitsToSubtract, to: date)!
```

Comment:
No `!` operator here please. Let's either throw an error or make the function return an optional